### PR TITLE
🌱 replace kubebuilder-release-tools with new workflow

### LINKED ISSUE #10925: Replace kubebuilder-release-tools action
Sooner or later the kubebuilder-release-tools actions will have to either migrate to the community-registry or be deprecated (https://github.com/kubernetes-sigs/kubebuilder-release-tools/issues/60#issuecomment-2236068489).

While thinking about this, I was wondering if we actually need the kubebuilder-release-tools action in Cluster API.
Especially considering how much effort it took the last time to get a change implemented in kubebuilder-release-tools and the action correctly published and bumped.

I took a closer look and it looks like we use it only to verify PR titles in the following way:
* to check for icons (while ignoring e.g. WIP): https://github.com/kubernetes-sigs/kubebuilder-release-tools/blob/master/verify/type.go
* to check if there is an issue referenced in the PR title: https://github.com/kubernetes-sigs/kubebuilder-release-tools/blob/master/verify/issue.go

I think we can replace the usage of kubebuilder-release-tools with a small GitHub action that just runs some local code (e.g. bash) to achieve the same goal.

Once implemented, let's make sure we surface this alternative implementation at the kubebuilder-release-tools issue (https://github.com/kubernetes-sigs/kubebuilder-release-tools/issues/60)

### PR DESCRIPTION
**Pull request by sivchari**
**What this PR does / why we need it**:

I replace kubebuilder-release-tools with self-workflow.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10925

### CODE REVIEW

**Comment by sivchari** [comment]
I'm ready for review.

File: .github/workflows/pr-verify.yaml [hunk 1]
```diff
@@ -5,15 +5,64 @@ on:
     types: [opened, edited, synchronize, reopened]
 
 permissions:
-  checks: write # Allow access to checks to write check runs.
+  contents: read
+  pull-requests: write
```
Comment (sbueringer) [inline .github/workflows/pr-verify.yaml:9]: I would prefer if we only log those messages in the actions and not write to the PR

I think adding the comments to the PR will lead to a lot of noise

(Prow usually implements some cleanup to delete comments after issues are fixed, but I don't want to do that here)

Also just a lot better if we require less permissions for security reasons

File: .github/workflows/pr-verify.yaml [hunk 2]
```diff
@@ -5,15 +5,64 @@ on:
...
+
+env:
+  GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
 jobs:
-  verify:
+  verify-pr-type:
```
Comment (sbueringer) [inline .github/workflows/pr-verify.yaml:15]: Let's combine the two jobs into just one (I assume two jobs lead to two "checks" on the PR, less checks is better here)

File: .github/workflows/pr-verify.yaml [hunk 3]
```diff
@@ -5,15 +5,64 @@ on:
...
-    - name: Verifier action
-      id: verifier
-      uses: kubernetes-sigs/kubebuilder-release-tools@012269a88fa4c034a0acf1ba84c26b195c0dbab4 # tag=v0.4.3
-      with:
-        github_token: ${{ secrets.GITHUB_TOKEN }}
+      - uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # tag=v4.1.7
+
+      - name: Check if PR title has acceptable prefix
+        run:  |
+          targets=(":sparkles:" ":bug:" ":book:" ":seedling:" ":warning:" ":rocket:" ":running:" "✨" "🐛" "📚" "🌱" "⚠️" "🚀" "🏃")
```
Comment (sbueringer) [inline .github/workflows/pr-verify.yaml:23]: Sorry, slightly in contrast to what I wrote before. Can we move the implementation to a shell script in `hack/verify-pr-title.sh` please?
(the shell script should then just take the PR tilte as an argument)

This just makes it a lot easier to test the shell script manually a bit (and we get shellcheck linting for free)

Comment (sbueringer) [reply .github/workflows/pr-verify.yaml:23]: Very recently this scripped got added to kubebuilder: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/pr-title-checker.sh

Let's take a look if it makes sense to reuse it (please check if it covers all the checks we want)

File: .github/workflows/pr-verify.yaml (hunk 1 above)
Comment (sivchari) [reply .github/workflows/pr-verify.yaml:9]: Okay, I make it not to write to the PR.

File: .github/workflows/pr-verify.yaml (hunk 3 above)
Comment (sivchari) [reply .github/workflows/pr-verify.yaml:22]: Thanks for good comment. The script that you mentioned only check if the title contains valid prefix.
So I also check the new script that checks the title doesn't contain the number. 👍

**Comment by sivchari** [comment]
@sbueringer 
I move the workflow into script then run it.
Actually, it was so easy to check if the script is valid. It's godd approach we want.

File: .github/workflows/pr-verify.yaml [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
     steps:
-    - name: Verifier action
-      id: verifier
-      uses: kubernetes-sigs/kubebuilder-release-tools@012269a88fa4c034a0acf1ba84c26b195c0dbab4 # tag=v0.4.3
-      with:
-        github_token: ${{ secrets.GITHUB_TOKEN }}
+      - uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # tag=v4.1.7
+
+      - name: Check if PR title is valid
+        run: sh hack/verify-pr-title.sh ${{ github.event.pull_request.title }}
```
Comment (chrischdi) [inline .github/workflows/pr-verify.yaml:18]:
```suggestion
        run: sh hack/verify-pr-title.sh "${{ github.event.pull_request.title }}"
```

I think we need quote this?

File: hack/verify-pr-title.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+# Normalize common emojis in text form to actual emojis
+trimmed_title=$(echo "$trimmed_title" | sed -E "s/:warning:/⚠/g")
+trimmed_title=$(echo "$trimmed_title" | sed -E "s/:sparkles:/✨/g")
+trimmed_title=$(echo "$trimmed_title" | sed -E "s/:bug:/🐛/g")
+trimmed_title=$(echo "$trimmed_title" | sed -E "s/:book:/📖/g")
+trimmed_title=$(echo "$trimmed_title" | sed -E "s/:rocket:/🚀/g")
+trimmed_title=$(echo "$trimmed_title" | sed -E "s/:seedling:/🌱/g")
+
+# Check PR type prefix
+if [[ "$trimmed_title" =~ ^⚠ ]] || [[ "$trimmed_title" =~ ^✨ ]] || [[ "$trimmed_title" =~ ^🐛 ]] || [[ "$trimmed_title" =~ ^📖 ]] || [[ "$trimmed_title" =~ ^🚀 ]] || [[ "$trimmed_title" =~ ^🌱 ]]; then
```
Comment (chrischdi) [inline hack/verify-pr-title.sh:34]:
```suggestion
if [[ "$trimmed_title" =~ ^(⚠|✨|🐛📖|🚀|🌱) ]]; then
```
Does this also work?

File: .github/workflows/pr-verify.yaml (hunk 4 above)
Comment (sivchari) [reply .github/workflows/pr-verify.yaml:18]: Yes, we need quote.

**Comment by sivchari** [comment]
Fixed, thanks.

**Review by chrischdi — COMMENTED** [review]
/lgtm

File: .github/workflows/pr-verify.yaml [hunk 6]
```diff
@@ -1,19 +1,19 @@
...
+name: PR title verifier
 
 on:
   pull_request_target:
     types: [opened, edited, synchronize, reopened]
 
-permissions:
-  checks: write # Allow access to checks to write check runs.
+env:
+  GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (sbueringer) [inline .github/workflows/pr-verify.yaml:8]: Do we need this?

We have many other actions using checkout, but none of them needs this

**Review by sbueringer — COMMENTED** [review]
Script looks good to me. Just one other nit

**Comment by sbueringer** [comment]
We should cherry-pick this everywhere once it's merged

File: .github/workflows/pr-verify.yaml (hunk 6 above)
Comment (sivchari) [reply .github/workflows/pr-verify.yaml:8]: Oh, sorry. This was used to write to PR, but forgot to erase it. Thanks

Comment (sbueringer) [reply .github/workflows/pr-verify.yaml]: No problem

**Comment by sbueringer** [comment]
Thank you very much!!

/lgtm
/approve

/hold
Let's merge after the patch releases today

**Comment by sbueringer** [comment]
@chrischdi Note for us (I'll keep track of it). Once we verified this a bit in core CAPI I would like to do the same change in CAPV, CR & CT (+ cherry-pick back in all repos, so supported branches still work even if the kubebuilder image becomes unavailable eventually)

**Comment by sivchari** [comment]
@sbueringer 
Should we share the information with kubebuilder team ? Thanks.

**Comment by sbueringer** [comment]
> @sbueringer Should we share the information with kubebuilder team ? Thanks.

That we extended their script? Sure, feel free to open an issue in the kubebuilder repo to suggest improvements

**Comment by sbueringer** [comment]
Hm doens't work: https://github.com/kubernetes-sigs/cluster-api/actions/runs/10805038279/job/29971485167?pr=10959

Did we test this on Linux?

**Comment by sivchari** [comment]
I only ran this on mac. Sorry. I'll attempt this on linux and fix until today.

**Comment by chrischdi** [comment]
its a sh vs bash thing. We run it using `sh`:

https://github.com/kubernetes-sigs/cluster-api/blob/main/.github/workflows/pr-verify.yaml#L15

I'll take a look

**Comment by sbueringer** [comment]
No worries, thx for the fix!!

**Comment by sivchari** [comment]
thanks for all!

**Pull request merged**